src/TheSpace/HarbergerMarket.sol

```diff
@@ -1,59 +1,74 @@
-pragma solidity ^0.8.0;
+pragma solidity ^0.8.11;
 
 import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
 import "@openzeppelin/contracts/utils/Multicall.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
 
 import "./Property.sol";
 
 /**
  * @dev Market place with Harberger tax, inherits from `IPixelCanvas`. Market creates one ERC721 contract as property, and attaches one ERC20 contract as currency.
  */
-abstract contract HarbergerMarket is Multicall {
+abstract contract HarbergerMarket is Multicall, Ownable {
+    error PriceTooLow();
+    error Unauthorized();
+    error TokenNotExists();
+
     /**
-     * @dev Emitted when a token changes price.
+     * @dev Tax record of token. Use block number to record tax collection time.
+     *
+     * TODO: more efficient storage scheme, see: https://medium.com/@novablitz/storing-structs-is-costing-you-gas-774da988895e
      */
-    event Price(uint256 indexed tokenId, uint256 price);
+    struct TaxRecord {
+        uint256 price;
+        uint256 lastTaxCollection;
+        uint256 ubiWithdrawn;
+    }
 
     /**
-     * @dev Emitted when tax is collected.
+     * @dev Tax configuration of market.
+     * - taxRate: Tax rate in bps every 1000 blocks
+     * - treasuryShare: Share to treasury in bps.
      */
-    event Tax(uint256 indexed tokenId, uint256 amount);
+    enum ConfigOptions {
+        taxRate,
+        treasuryShare
+    }
 
     /**
-     * @dev Emitted when UBI is distributed.
+     * @dev Emitted when a token changes price.
      */
-    event UBI(uint256 indexed tokenId, uint256 amount);
+    event Price(uint256 indexed tokenId, uint256 price, address owner);
 
     /**
-     * @dev Tax record of token.
-     *
-     * TODO: more efficient storage scheme, see: https://medium.com/@novablitz/storing-structs-is-costing-you-gas-774da988895e
+     * @dev Emitted when tax configuration updates.
      */
-    struct TaxRecord {
-        uint256 price;
-        uint256 lastTaxCollection;
-        uint256 ubiWithdrawn;
-    }
+    event Config(ConfigOptions indexed option, uint256 value);
 
     /**
-     * @dev Mapping from token id to tax record.
+     * @dev Emitted when tax is collected.
      */
-    mapping(uint256 => TaxRecord) public taxRecord;
+    event Tax(uint256 indexed tokenId, uint256 amount);
 
     /**
-     * @dev Tax rate in percentage.
+     * @dev Emitted when UBI is distributed.
      */
-    uint256 public taxRate;
+    event UBI(uint256 indexed tokenId, uint256 amount);
 
     /**
-     * @dev Share for community treasury in percentage.
+     * @dev Tax record of each token.
      */
-    uint256 public treasuryShare;
+    mapping(uint256 => TaxRecord) public taxRecord;
+
+    // Setting for tax config
+    mapping(ConfigOptions => uint256) public taxConfig;
 
+    // total accumulated UBI
     uint256 public accumulatedUBI;
 
-    uint256 public totalSupply;
+    // total token supply
+    uint256 public totalSupply = 1000000;
```

Comment:
why `taxConfig` is a mapping? looks we can pack these variables (`taxRate`, `treasuryShare`, `totalSupply`) into one slot.

```solidity
uint64 public taxRate;
uint64 public treasuryShare
uint128 public totalSupply
```

Reply:
With mapping, we can have a general `setTaxConfig` function to update the config. Are there ways to achieve similar effect with struct?